GameServerLib/Items/Shop.cs

```diff
@@ -74,13 +71,22 @@ public bool HandleItemBuyRequest(int itemId)
             return true;
         }
 
-        private void RemoveItem(IItem item, byte slotId, byte stackSize = 0)
+        private void RemoveItem(IItem item, byte slotId, byte stacksToRemove = 1)
         {
             var inventory = _owner.Inventory;
-            _owner.Stats.RemoveModifier(item.ItemData);
-            _game.PacketNotifier.NotifyRemoveItem(_owner, slotId, stackSize);
-            _owner.RemoveSpell((byte)(slotId + ITEM_ACTIVE_OFFSET));
-            inventory.RemoveItem(item);
+            var newStacks = (byte)Math.Max(item.StackCount - stacksToRemove, 0);
+
+            if(item.WithinBounds(newStacks))
+                item.SetStacks(newStacks);
```

Comment:
Mixed responsibilities. If you want to decrease the stack count from an item, you should do it with function from item class. I think that `IncrementStackCount` and alike are still important methods as it saves you checks for all `IStackable`. Usage of these methods like:
```
            if (i.ItemData.MaxStack > 1)
            {
                i.DecrementStackCount();
            }
```
should be avoided, as all the checks handled in the `IStackable` interface.

I suggest to simply remove this function as it's just causes confusion. Copy-paste it to the `HandleItemSellRequest`.

Reply:
Yeah, most of the mixed responsibility issues are regression because of the original broken `RemoveItem` method.

I'll revert most these changes and just leave the original (fixed) implementation in the `HandleItemSellRequest` method instead.